src/schema_compile.c

```diff
@@ -1199,6 +1197,43 @@ lys_compile_unres_llist_dflts(struct lysc_ctx *ctx, struct lysc_node_leaflist *l
     return LY_SUCCESS;
 }
 
+/**
+ * @brief Iteratively get all leafrefs from @p node
+ * if the node is of type union, otherwise just return the leafref.
+ *
+ * @param[in] node Node that may contain the leafref.
+ * @param[in,out] index Value that is passed between function calls.
+ * For each new node, initialize value of the @p index to 0, otherwise
+ * do not modify the value between calls.
+ * @return Pointer to the leafref or next leafref, otherwise NULL.
+ */
+static struct lysc_type_leafref *
+lys_type_leafref_next(const struct lysc_node *node, uint64_t *index)
+{
+    uint64_t array_count;
+    struct lysc_type *leaf_type, *union_type;
+    struct lysc_type_leafref *ret = NULL;
+
+    leaf_type = ((struct lysc_node_leaf *)node)->type;
+    if (leaf_type->basetype == LY_TYPE_UNION) {
+        array_count = LY_ARRAY_COUNT(((struct lysc_type_union *)leaf_type)->types);
+        for (uint64_t i = *index; !ret && (i < array_count); i++) {
+            union_type = ((struct lysc_type_union *)leaf_type)->types[i];
+            if (union_type->basetype == LY_TYPE_LEAFREF) {
+                *index = i + 1;
+                ret = (struct lysc_type_leafref *)union_type;
+            }
+        }
```

Comment:
It still seems redundantly complex to me. There are probably more ways of improving it.